ca/certificate-authority.go

```diff
@@ -172,8 +184,68 @@ func loadKey(keyConfig cmd.KeyConfig) (priv crypto.Signer, err error) {
 	return
 }
 
+// checkHSMFault checks whether there has been an HSM fault observed within the
+// timeout window.  CA methods that use the HSM should call this method right
+// away, to minimize the performance impact of HSM outages.
+func (ca *CertificateAuthorityImpl) checkHSMFault() bool {
+	ca.hsmFaultLock.Lock()
+	defer ca.hsmFaultLock.Unlock()
+
+	now := ca.Clk.Now()
+	timeout := ca.hsmFaultLastObserved.Add(ca.hsmFaultTimeout)
+	if timeout.Before(now) {
+		// Go ahead and try again
+		return false
+	}
+
+	return ca.hsmFault
+}
+
+// noteHSMFault updates the CA's state with regard to HSM faults.  CA methods
+// that use an HSM should pass errors that might be HSM errors to this method.
+// We distinguish HSM errors from other errors by looking for a string that the
+// PKCS#11 library appends to its errors.
+//
+// XXX(rlb): It would be better to directly check that the error is of the
+// PKCS#11 library's local Error type.  But that would be messier, since we
+// would have to import that library, which we don't now.
+func (ca *CertificateAuthorityImpl) noteHSMFault(err error) {
+	ca.hsmFaultLock.Lock()
+	defer ca.hsmFaultLock.Unlock()
+
+	// If err == nil, the HSM is working, reset the timeout
```

Comment:
If we're never going to impreove the timeout, then we should just just have a fixed delay.  I choose... 1min.